pkg/manager/member/tikv_scaler_test.go

```diff
@@ -543,90 +906,1150 @@ func TestTiKVScalerScaleIn(t *testing.T) {
+		newSet := oldSet.DeepCopy()
+		newSet.Spec.Replicas = pointer.Int32Ptr(3)
+		helper.SetDeleteSlots(newSet, test.newDeleteSlots)
+
+		createPodFn := func(ordinal int, storeIdLabel string) {
+			pod := &corev1.Pod{
+				TypeMeta: metav1.TypeMeta{Kind: "Pod", APIVersion: "v1"},
+				ObjectMeta: metav1.ObjectMeta{
+					Name:              TikvPodName(tc.GetName(), int32(ordinal)),
+					Namespace:         corev1.NamespaceDefault,
+					CreationTimestamp: metav1.Time{Time: time.Now().Add(-1 * time.Hour)},
+				},
+			}
+			readyPodFunc(pod)
+			pvc1 := _newPVCForStatefulSet(oldSet, v1alpha1.TiKVMemberType, tc.Name, int32(ordinal))
+			pvc2 := pvc1.DeepCopy()
+			pvc1.Name = pvc1.Name + "-1"
+			pvc1.UID = pvc1.UID + "-1"
+			pvc2.Name = pvc2.Name + "-2"
+			pvc2.UID = pvc2.UID + "-2"
+			pvcIndexer.Add(pvc1)
+			pvcIndexer.Add(pvc2)
+			pod.Spec.Volumes = append(pod.Spec.Volumes,
+				corev1.Volume{
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc1.Name,
+						},
+					},
+				}, corev1.Volume{
+					VolumeSource: corev1.VolumeSource{
+						PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+							ClaimName: pvc2.Name,
+						},
+					},
+				})
+
+			pod.Labels = map[string]string{}
+			pod.Labels[label.StoreIDLabelKey] = storeIdLabel
+			podIndexer.Add(pod)
+		}
+
+		createPodFn(0, "10")
+		createPodFn(1, "11")
+		createPodFn(2, "12")
+		createPodFn(3, "13")
+		createPodFn(4, "1")
+
+		pdClient := controller.NewFakePDClient(pdControl, tc)
+		pdClient.AddReaction(pdapi.GetConfigActionType, func(action *pdapi.Action) (interface{}, error) {
+			var replicas uint64 = 3
+			return &pdapi.PDConfigFromAPI{
+				Replication: &pdapi.PDReplicationConfig{
+					MaxReplicas: &replicas,
+				},
+			}, nil
+		})
+		if test.getStoresFn == nil {
+			test.getStoresFn = func(action *pdapi.Action) (interface{}, error) {
+				store := &pdapi.StoreInfo{
+					Store: &pdapi.MetaStore{
+						StateName: v1alpha1.TiKVStateUp,
+						Store: &metapb.Store{
+							Address: fmt.Sprintf("%s-tikv-0", "basic"),
+						},
+					},
+				}
+				return &pdapi.StoresInfo{
+					Count:  5,
+					Stores: []*pdapi.StoreInfo{store, store, store, store, store},
+				}, nil
+			}
+		}
+		pdClient.AddReaction(pdapi.GetStoresActionType, test.getStoresFn)
+
+		features.DefaultFeatureGate.Set(fmt.Sprintf("AdvancedStatefulSet=%v", test.enableAsts))
+		actualSet := oldSet.DeepCopy()
+		for _, op := range test.ops {
+			if op.preHandler != nil {
+				op.preHandler(tc)
+			}
+			desiredSet := newSet.DeepCopy()
+			_ = scaler.ScaleIn(tc, actualSet, desiredSet)
+			if diff := cmp.Diff(op.replicas, *desiredSet.Spec.Replicas); diff != "" {
+				t.Errorf("unexpected (-want, +got): %s", diff)
+			}
+			if diff := cmp.Diff(op.deleteSlots, helper.GetDeleteSlots(desiredSet)); diff != "" {
+				t.Errorf("unexpected (-want, +got): %s", diff)
+			}
+			actualSet = desiredSet.DeepCopy()
```

Comment:
do we need to check the final state of `actualSet` after the `for` loop?

Reply:
I think check `replica` and `deleteSlots` is enough, since `Scale` function only change these two fields. Same logic with the former test like `TestTiKVScalerScaleIn`.